### DISCUSSION TITLE
generate schema dynamically?

### DISCUSSION THREAD

**Discussion by willvlad** [vercel/next.js#21696] [Help]
Been trying to generate a schema dynamically based on the blog title. 

```
useEffect(() => {
   
    if (props.data) {
      if (utils.undefinedCheck(props.data.title)) {
      
        setSchema({
          "@context": "https://schema.org",
          "@type": "NewsArticle",
          headline: props.data.title
        });
      }
    }
  }, []);
```

The server rendered pages however don't have the title as per the schema above. Anyone know what's causing the issue? Looks like useEffect is not having any effect on updating the title and schema for some reason...

**Comment by gubareve** [comment]
You may want to put `props.data.title` in the 2nd argument in useEffect (`useEffect(..., [props.data.title])`) this makes the title update every time the title changes. This means that if it is originally rendered with props.data returning null, and then updating to return the title,  useEffect will re-run.

**Reply by willvlad** [reply]
Thanks, but it shouldn't affect that behaviour as [] should re-render on load too.

The problem seems to be that a schema in Nextjs gets generated before a component had a chance to fully load and doesn't re-creates itself after that so it's not a reliable way to manage it dynamically. 

My workaround is not to use lifecycle methods at all, but take the url from router and set title, etc from props.

**Comment by Sangdt** [comment]
You should try [next-seo](https://github.com/garmeeh/next-seo) for this, they have a specific section for schema and their seo support is very good too.

**Reply by willvlad** [reply]
Good idea, i'll try that

**Comment by alexcloudstar** [comment]
Any updates on that? Is next-seo working properly?
